### PR TITLE
Improving reading and recreating fit pseudodata

### LINKED ISSUE #1323: should pseudata.py should use `fitting_data_dict`?
I'm asking because `pseudata.py` is using `common_data_reader` https://github.com/NNPDF/nnpdf/blob/74d91831393fff3335397037e498ea7bfd88e492/validphys2/src/validphys/pseudodata.py#L332 which is not what n3fit uses any more.

(@siranipour pinging you since I noticed that while looking at #1285, but it's clearly not part of that PR so didn't want to add entropy there)

### PR DESCRIPTION
**Pull request by siranipour**
Now that python is used to construct the pseudodata, I've updated the pseudodata.py function that regenerates the pseudodata of a given fit. It's now much easier to do this in parallel because we don't use swig objects anymore.

Closes #1323

The tests will, however, fail for now until I regenerate a PDF that saves its python pseudodata for the regression test. @wilsonmr, I may be misremembering but did you at some point have a PR which saved the pseudodata of a fit?

**EDIT**: This has turned into a bit of a general improvements PR.
1. _Corrected_ the recreation of a fit pseudodata. I've swapped the C++ make_replica with the python one.
2. _Improved_ the reading of pseudodata from a fit which has `fitting::savepseudodata: True`
3. There was a function that created the tr_mask for all replicas:
https://github.com/NNPDF/nnpdf/blob/b636ae04e55264efe9075cd7442fc2844293794d/validphys2/src/validphys/n3fit_data.py#L386
I've now changed this so it works on an individual replica and then `collect`ed it over `replicas` (I did this because it was useful for the recreation of a training validation mask.
4. Fixed the tests

### CODE REVIEW

**Comment by siranipour** [comment]
@scarlehoff this PR flies in tandem with #1333, in that it reproduces the pseudodata _a posteriori_ which is in principle the ones that are saved by #1333. I would appreciate if you could take a look at this soon too.

The basic idea is that I've added actions that work on a per replica basis. I.e replicate the pseudodata of `replica_i`. We then collect over either `NSList(range(nrep), nskey='replica')` to replicate all replicas, or `produce_pdfreplicas`, which is basically the same thing but accounting for the postfit reshuffling

File: validphys2/src/validphys/tests/test_pseudodata.py [hunk 1]
```diff
@@ -2,111 +2,81 @@
...
-)
-def setup_dicts(request):
-    n_process_config = dict(NPROC=request.param)
-    exp_infos_bytes = read_binary(validphys.tests.regressions, "test_exp_infos.pickle")
-    ns = yaml.safe_load(EXAMPLE_RUNCARD)
-    # This is what all the fitted replicas saw
-    exp_infos = pickle.loads(exp_infos_bytes)
+  pdf = API.pdf(pdf=PSEUDODATA_FIT)
+  # -1 because we ignore replica 0
+  assert len(pdf) - 1 == len(pdf_pseudodata)
```
Comment (scarlehoff) [inline validphys2/src/validphys/tests/test_pseudodata.py:35]: Please change the indentation to 4 spaces.

Comment (siranipour) [reply validphys2/src/validphys/tests/test_pseudodata.py:35]: Good spot, I've now reformatted this entire file using black

File: validphys2/src/validphys/n3fit_data.py [hunk 2]
```diff
@@ -405,29 +405,77 @@ def validation_pseudodata(pseudodata_table, training_mask):
...
+    ])
+    return pd.DataFrame(
+        all_masks,
+        columns=[f"replica {replica}"],
+        index=experiments_index
+    )
+
+replicas_training_mask_table = collect("replica_training_mask_table", ("replicas",))
+@table
+def training_mask_table(replicas_training_mask_table):
```
[the remark this answers is not in the document]
Comment (siranipour) [reply validphys2/src/validphys/n3fit_data.py:468]: Sorry I had forgotten to delete the old function

File: validphys2/src/validphys/pseudodata.py [hunk 3]
```diff
@@ -55,64 +50,50 @@ def read_fit_pseudodata(fitcontext, context_index):
     Example
     -------
     >>> from validphys.api import API
-    >>> data_indices_list = API.read_fit_pseudodata(fit="NNPDF31_nnlo_as_0118_DISonly_pseudodata", use_cuts="fromfit")
+    >>> data_indices_list = API.read_fit_pseudodata(fit="pseudodata_test_fit_n3fit")
     >>> len(data_indices_list) # Same as nrep
-    100
+    5
```
[the remark this answers is not in the document]
Comment (siranipour) [reply validphys2/src/validphys/pseudodata.py:55]: Good spot

File: validphys2/src/validphys/config.py [hunk 4]
```diff
@@ -237,6 +238,51 @@ def parse_fit(self, fit: str):
...
+        replica.
+        """
+        pdf = fitpdf['pdf']
+        replicas = fitted_replica_indexes(pdf)
+        return NSList(replicas, nskey='replica')
+
+    def produce_fitenvironment(self, fit, fitinputcontext):
+        """Like fitcontext, but additionally forcing various other
+        parameters, such as the cuts policy and Monte Carlo seeding to be
+        the same as the fit.
```
[the remark this answers is not in the document]
Comment (siranipour) [reply validphys2/src/validphys/config.py:259]: In this cases these will be unused keys right, this is unfortunately something reportengine doesn't flag cc @Zaharid

[the remark this answers is not in the document]
Comment (scarlehoff) [reply validphys2/src/validphys/config.py:259]: They will be unused but the user doesn't necessarily know that. Even if it cannot be truly caught, maybe a warning like the one for q2min would help:

```
[WARNING]: Using q2min from runcard
[WARNING]: Using w2min from runcard
```

**Review by scarlehoff — APPROVED** [review]

File: validphys2/src/validphys/config.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (siranipour) [reply validphys2/src/validphys/config.py:259]: Good idea, that is done now

**Pull request merged**